WIP: Add ChainRules

AlexRobson opened the pull request:
Adds support for ChainRules 1.0. 

This tests
```
*(::AbstractVecOrMat, ::Woodbury)::Matrix
*(::Real, ::Woodbury)::WoodburyPDMat
```

and adds in some constructor tests too.

AlexRobson commented on Project.toml:
```diff
@@ -1,21 +1,24 @@
 [deps]
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"
+ChainRulesTestUtils = "cdddcdb0-9152-4a09-a978-84456f9df70a"
+FiniteDifferences = "26cc04aa-876d-5657-8c51-4c34ba976000"
```
testdeps added - need moving

AlexRobson commented on Manifest.toml:
```diff
@@ -0,0 +1,306 @@
+# This file is machine-generated - editing it directly is not advised
```
neeeds removing

AlexRobson commented on test/woodbury_pd_mat.jl:
```diff
@@ -45,22 +45,19 @@
         @test c * W_dense ≈ c * W atol=1e-6
         @test (c * W) isa WoodburyPDMat
 
-        c = Diagonal(2.0 * ones(4,))
```
These changes are related to #18

mzgubic commented on src/woodbury_pd_mat.jl:
```diff
@@ -90,4 +88,6 @@ end
-*(c::Real, a::WoodburyPDMat) = a * c
\ No newline at end of file
+*(c::Real, a::WoodburyPDMat) = a * c
+*(c::Diagonal{T}, a::WoodburyPDMat) where {T<:Real} = c * Matrix(a)
```
these two are not needed, right?
```julia
julia> A = randn(4, 2); # on master

julia> D = Diagonal(randn(2).^2 .+ 1);

julia> S = Diagonal(randn(4).^2 .+ 1);

julia> d = Diagonal(rand(4));

julia> W = WoodburyPDMat(A, D, S);

julia> d*W == d*Matrix(W)
true

julia> W*d == Matrix(W)*d
true
```

AlexRobson replied:
Oh you are correct.

AlexRobson replied:
Oh not in the case of the Real - just using the abstractmatrix fallback will cause it to densify.

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,40 @@
+@non_differentiable validate_woodbury_arguments(A, D, S)
+
+function ChainRulesCore.rrule(
+    ::typeof(*), A::Real, B::WoodburyPDMat{T, TA, TD, TS}
+) where {T, TA, TD, TS}
```
```suggestion
function ChainRulesCore.rrule(::typeof(*), A::Real, B::WoodburyPDMat)
```

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,40 @@
+        )
+    end
+
+    function times_pullback(ȳ::Tangent{<:WoodburyPDMat})
+        Ȳ = unthunk(ȳ)
```
```suggestion
    function times_pullback(Ȳ::Tangent{<:WoodburyPDMat})
```
We already know it's a tangent, not thunk. We should probably add a pullback dispatch which just unthunks. This broke type inference before, and the fix was to define the pullback methods outside, see something like 
https://github.com/JuliaDiff/ChainRules.jl/blob/e6a27db89cbb392a1f55fba7bae19caebd9fe43a/src/rulesets/LinearAlgebra/structured.jl#L107-L114

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,40 @@
+) where {T, TA, TD, TS}
+    project_A = ProjectTo(A)
+    project_B = ProjectTo(B)
+    function times_pullback(ȳ::AbstractMatrix)
+        Ȳ = unthunk(ȳ)
+        Ā = dot(Ȳ, B)
+        B̄ = A' * Ȳ
+        return (
+           NoTangent(),
+           @thunk(project_A(Ā')),
+           @thunk(project_B(B̄)),
+        )
+    end
```
```suggestion
    function times_pullback(ȳ::AbstractMatrix)
        Ā = @thunk(project_A(dot(Ȳ, B)'))
        B̄ = @thunk(project_B(A' * Ȳ))
        return (NoTangent(), Ā, B̄)
    end
```
we already know it is not a thunk from the pullback argument type

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,40 @@
+end
+
+function ChainRulesCore.ProjectTo(W::WoodburyPDMat)
+    function dW(W̄)
```
Is it not possible to construct `ProjectTo` object? And then make it callable? All the examples in the ChainRulesCore are built this way, see https://github.com/JuliaDiff/ChainRulesCore.jl/blob/master/src/projection.jl.

We have decided against closures because they are not easily extensible. I.e. if someone defines their own type and wants to make sure it can be projected to a Woodbury then they must either pirate or change this function (and introduce a dependency to PDMatsExtras.

AlexRobson replied:
Oh right. Will change.

AlexRobson replied:
updated

mzgubic commented on test/chainrules.jl:
```diff
@@ -0,0 +1,76 @@
+# We do this because the Woodbury cannot represent it's own tangent. 
+# I.e. for Y = f(W,...), W̄ = ∂Y / ∂W, is not necessarily a valid Woodbury. 
+# Consider, e.g. the case of a Positive Diagonal matrix
+struct WoodburyLike
```
I may not understand the problem enough, but could we just add a `rand_tangent` dispatch for the Woodbury which makes sure that the matrix is PSD?

mzgubic commented on test/chainrules.jl:
```diff
@@ -0,0 +1,76 @@
+        test_rrule(*, rand(4,4), W)
+    end
+
+    @testset "*(Real-Woodbury" begin
```
```suggestion
    @testset "*(Real, Woodbury)" begin
```

mzgubic commented:
what's the error for 2. ? I will try to take a look at the end of the day

AlexRobson commented on test/woodbury_pd_mat.jl:
```diff
@@ -48,6 +48,14 @@
         @test c * W_dense ≈ c * W atol=1e-6
         @test (c * W) isa WoodburyPDMat
+
+        c = Diagonal(rand(4,))
```
Agree - added them because i readded the algebra above.

AlexRobson commented:
> what's the error for 2. ? I will try to take a look at the end of the day

It is numerically very very wrong.

mzgubic commented:
Not sure why the passing the Tangent was failing (the math looked fine) but solved it by creating an intermediate Woodbury. Also putting the `times_pullback` outside of the `rrule` somehow solves the inference (we have seen this before, but I don't understand it).

I still don't know why the following fails though:
```julia
julia> W = WoodburyPDMat(rand(3,2), Diagonal(rand(2,)), Diagonal(rand(3,)))
3×3 WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}:
 1.26364   0.194865   0.14236
 0.194865  0.553561   0.0738048
 0.14236   0.0738048  0.366308

julia> T = Tangent{WoodburyPDMat}(;A=W.A, D=W.D, S=W.S)
Tangent{WoodburyPDMat}(A = [0.6529299263287578 0.9705937318111584; 0.2742328923999924 0.5824550169874951; 0.14327303956778725 0.6355464664765937], D = [0.6900525901684949 0.0; 0.0 0.12613565656887138], S = [0.8506296051551665 0.0 0.0; 0.0 0.4588742330641957 0.0; 0.0 0.0 0.3011946555949865])

julia> test_rrule(*, 2.0, W ⊢ W; output_tangent=W)
test_rrule: * on Float64,WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}: Test Failed at /Users/mzgubic/.julia/packages/ChainRulesTestUtils/8380y/src/check_result.jl:24
  Expression: isapprox(actual, expected; kwargs...)
   Evaluated: isapprox(2.1647628223399655, 1.5169372885519636; rtol = 1.0e-9, atol = 1.0e-9)
Stacktrace:
 [1] test_approx(actual::Float64, expected::Float64, msg::String; kwargs::Base.Iterators.Pairs{Symbol, Float64, Tuple{Symbol, Symbol}, NamedTuple{(:rtol, :atol), Tuple{Float64, Float64}}})
   @ ChainRulesTestUtils ~/.julia/packages/ChainRulesTestUtils/8380y/src/check_result.jl:24
 [2] macro expansion
   @ ~/.julia/packages/ChainRulesTestUtils/8380y/src/testers.jl:238 [inlined]
 [3] macro expansion
   @ /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
 [4] test_rrule(::ChainRulesTestUtils.ADviaRuleConfig, ::typeof(*), ::Float64, ::Vararg{Any, N} where N; output_tangent::WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}, check_thunked_output_tangent::Bool, fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, rrule_f::Function, check_inferred::Bool, fkwargs::NamedTuple{(), Tuple{}}, rtol::Float64, atol::Float64, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ ChainRulesTestUtils ~/.julia/packages/ChainRulesTestUtils/8380y/src/testers.jl:194
Test Summary:                                                                                                                            | Pass  Fail  Total
test_rrule: * on Float64,WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}} |    8     1      9
ERROR: Some tests did not pass: 8 passed, 1 failed, 0 errored, 0 broken.
```

Also, I don't think we have fully figured out our story in testing arbitrary tangents (like passing a Matrix instead of the Woodbury as `output_tangent`). I will write an issues about this.

AlexRobson commented on src/chainrules.jl:
```diff
@@ -0,0 +1,31 @@
+end
+_times_pullback(ȳ::AbstractThunk, A, B, proj) = _times_pullback(unthunk(ȳ), A, B, proj)
+function _times_pullback(Ȳ::Tangent{<:WoodburyPDMat}, A, B, proj)
+    W = WoodburyPDMat(Ȳ.A, Ȳ.D, Ȳ.S)
```
I still confused on why this is valid - does the tangent backing have to be able to reproduce the woodbury - i.e. considering the diagonals (which should be pos def) - their tangent is bound to positive? Again the PositiveReal float as an example.

mzgubic replied:
Not sure I follow - the pullback above would hit the `AbstractArray` method of `times_pullback`, right?

AlexRobson replied:
Yes, but the point is on the tangent that is produced: 

```
julia> ∂W = unthunk(ChainRulesCore.rrule(*, D, W)[2](rand_tangent(D * W))[3])
Tangent{WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}}(A = [-0.26343438999522384 -1.3809461745889253; 2.8422893418996336 -0.13508471772276032; -10.328690745141067 -5.416957153117603; 5.441344472486843 3.0253900899189343], D = [-1.2857824438446035 0.0; 0.0 -4.303274675114356], S = [0.18734959448406613 0.0 0.0 0.0; 0.0 3.094752052943939 0.0 0.0; 0.0 0.0 -2.9860594258216753 0.0; 0.0 0.0 0.0 -0.26627135393506735])

julia> WoodburyPDMat(∂W.A, ∂W.D, ∂S)
ERROR: UndefVarError: ∂S not defined
Stacktrace:
 [1] top-level scope
   @ REPL[66]:1

julia> WoodburyPDMat(∂W.A, ∂W.D, ∂W.S)
ERROR: ArgumentError: Detected negative element on diagonal of D: [-1.2857824438446035 0.0; 0.0 -4.303274675114356]
```

mzgubic replied:
Oh, is that after the projection is applied? I didn't quite figure out whether that is correct

AlexRobson replied:
There is afaict nothing wrong with the projection. This for example is passing (test_rrule(*, D, W)) and it is matching with the output of finite differences which doesn't use the projection. 

The point in question is again on the nature of the tangent. My understanding of your position is that it should always be valid to do `WoodburyPDMat(∂W.A, ∂W.D, ∂W.S)` and I don't understand this. My counterexample above is providing a tangent that would not fill this criteria. 

To try and fledge this out a bit, consider a type constrained to be positive real:

```
struct PositiveReal{T}
    x::T
    function PositiveReal(x::Real) = (@assert x > 0; new{T}(x))
end

and consider the pullback for this in any operation, say `*(A::Real, B::PositiveReal).` We have the pullbacks as `Ā = ΔΩ * B', B̄ = A' * ΔΩ`. If either A' or ΔΩ  are < 0  then B̄ < 0. This seems fine, as it's a gradient, but it does not mean that it can form a valid `PositiveReal`. Put another way the loss with respect to the element of B is negative, even though the primal is positive, so it doesn't make sense to try and do `PositiveReal{Tangent{PositiveReal}.x}`, which is effectively what `construct` in CR would try and do. To relate to my example above, I don't follow why it is necessary the case that `WoodburyPDMat(∂W.A, ∂W.D, ∂W.S)`  can be constructed (for example, with a negative element on the diagonals for the diagonals). 

Happy to accept if i've completely misunderstood something being less experienced in all this!

mzgubic replied:
Oh no you are absolutely right! I was just being really slow with understanding your point. Fleshing out the PositiveReal example cleared it up. I now understand your first point as well, sorry just didn't click the first time!

AlexRobson commented:
OK, I undestand a bit more about what's going on now:

Consider this snippet - (the tests pass at least locally (note I am using the commit prior to the one just added where the woodburyPDMat constructor is added to the pullback):

```
            primal = R * W

            # Generate the Tangent as ChainRulesTestUtils would do
            W̄ = rand_tangent(Random.GLOBAL_RNG, collect(primal))
            T = ProjectTo(primal)(W̄)
      
            res = ChainRulesCore.rrule(*, R, W)
            f_jvp = j′vp(ChainRulesTestUtils._fdm, x -> (*(x...)), T, (R, W))[1]

            # Expected
            R̄ = ProjectTo(R)(dot(W̄, W'))
            W̄ = ProjectTo(W)(conj(R) * W̄)

            R̄_rrule = unthunk(res[2](T)[2])
            W̄_rrule = unthunk(res[2](T)[3])

            @test res[1] == primal
            @test R̄ ≈ f_jvp[1]
            @test W̄.A ≈ f_jvp[2].A
            @test W̄.D ≈ f_jvp[2].D
            @test W̄.S ≈ f_jvp[2].S
```

ProjectTo is pushing the tangent W̄ into it's constituents, A, D and S. It is not the case that `W̄ = ∂W.A * ∂W.D * ∂W.A' + ∂W.S`. We want `W̄`.

AlexRobson commented:
I'm not sure what is intended is possible in the present set-up. If we consider the pullback and the projection:

```
function _times_pullback(Ȳ::AbstractMatrix, A, B, proj)
    Ā = proj.A(dot(Ȳ, B)')
    B̄ = proj.B(A' * Ȳ)
    return (NoTangent(), Ā, B̄)
end

function (W::ProjectTo{T})(W̄) where {T<:WoodburyPDMat}
    Ā(W̄) = ProjectTo(W.A)((W̄ + W̄') * (W.A * W.D))
    D̄(W̄) = ProjectTo(W.D)(W.A' * (W̄) * W.A)
    S̄(W̄) = ProjectTo(W.S)(W̄)
    return Tangent{T}(; A = Ā(W̄), D = D̄(W̄), S = S̄(W̄))
end

```

We want the pullback to support `Ȳ::Tangent{WoodburyPDMat`. The snippet above shows that passing through W̄ . 
However, this will have the fields A,D,S with their associated pullbacks due to the Projection. Not W̄ .

Two options to consider: 

We can't recover W̄ from the Projection (non of the entries are invertable, in general projections are likely to be lossy). Put another way, the structured differentials can't recover the natural differential (if this is the right language to use). Which may imply that this projection isn't what we want. 

We could extend the tangent representation to include the natural differential, however with some playing with the canonicalize this isn't possible as CRC assumes for any structured matrix the tangent fields match the primal fields. 

So we want the projection when testing the vectorised representation of the woodbury, but we also don't want it if we want to push the tangent back further down the chain rule. 

Though I've probably misunderstood something somewhere.

mzgubic commented:
I was thinking about the last option. I only think it is possible to go from Woodbury to a vector, but not back (as far as I know it is not possible to "factorise" an arbitrary dense matrix to a Woodbury?).

It might still solve our problems, because in some cases only going to a vector is required (and not going back).

AlexRobson commented:
I've cleaned up this MR now - it should be ready to go. 

As there were several discussion points previously, I'll give an overview:

* Updates to use ChainRulesCore v1. 
* Inplements a ProjectTo for the Woodbury that takes a dx::AbstractMatrix and projects that pushback onto the constituent elements of the Woodbury. Essentially, as we know that W = A D A' + S, we know how Ā, D̄, S̄ can be constructed from W̄. So if W̄ is a Matrix, Project is used to produce the `Tangent{WoodburyPDMat}(A = Ā, D = D̄, S = S̄)`. If W̄ is already a Tangent, Project just passes this through as it is in the correct subspace. 
* The rrules for Diagonal * Woodbury (and reverse), and the rrules for Real * Woodbury (and reverse) are implemented. These capture two different scenarios for the rrules. In the first case, abbreviated to `Dmat * W` this returns a dense Matrix in the primal, and thus the cotangent will be a Matrix. In the second case, abbreviated to R * W, this returns another Woodbury, and thus the cotangent will be of type `Tangent{WoodburyPDMat}`. So we have: 

  * (D * W)::Matrix - This is covered by existing rrules as the cotangent will be a Matrix and we have ProjectTo specified. 
  * (R * W)::WoodburyPDMat - This requires custom rrules because we need to work with a cotangent of type `Tangent{WoodburyPDMat}`. 
  
* We don't need to test (D * W) with a `Tangent{WoodburyPDMat}` cotangent because the primal type is a Matrix. I have added a  (R * W) with a `Matrix` cotangent test but we can't use `test_rrule` for this directly, as commented. 

* The WoodburyLike thing is removed. In retrospect I don't particularly like the Woodbury being force-constructed from the constiutent gradients, but it does mean that the existing tests can just be used (i.e. the output of `f_jvp` does create a Woodbury, but it's unlikely to be a valid one). 

Will take out of WIP.

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,63 @@
+
+# Rule for Woodbury * Real. 
+# Ignoring Complex version for now. 
+# NOTE: # Can't use Tangent{WoodburyPDMat} here. ChainRules is fine, but Zygote has issues and passes around Tangent{Any}
```
does this comment still apply? Does it refer to returning Tangent from the pullback, or dispatching on the argument of the pullback?

AlexRobson replied:
I didn't dig into this too much, but this is what I did. 

I was checking with an integration test with Zygote like the following:

```
@testset "Zygote Integration" begin
    m = randn(size(A, 1))
    α = 2.1
    test_ad(randn(), R, α, m, A, D, S, x) do R, α, m, A, D, S, x
        W = R * WoodburyPDMat(A, D, S)
        return logpdf(Distributions.GenericMvTDist(α, m, W), x)
    end
end
```

(this is entirely similar to what is going on [here](https://github.com/invenia/PDMatsExtras.jl/blob/master/test/woodbury_pd_mat.jl#L71). 

And basically, with the code-as-is this is fine. Without it, , i.e. with:

```
function _times_pullback(Ȳ::Tangent, A::T, B::Real, proj)  where {T<:WoodburyPDMat} ->
function _times_pullback(Ȳ::Tangent{WoodburyPDMat}, A::T, B::Real, proj)  where {T<:WoodburyPDMat}
```

I get: 
```
ERROR: MethodError: no method matching _times_pullback(::Tangent{Any, NamedTuple{(:A, :D, :S), Tuple{Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}}}, ::Float64, ::WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}, ::NamedTuple{(:A, :B), Tuple{ProjectTo{Float64, NamedTuple{(), Tuple{}}}, ProjectTo{WoodburyPDMat{Float64, Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}, NamedTuple{(:A, :D, :S), Tuple{Matrix{Float64}, Diagonal{Float64, Vector{Float64}}, Diagonal{Float64, Vector{Float64}}}}}}})
Closest candidates are:
  _times_pullback(::Tangent, ::Any, ::Any, ::Any, ::Any) at ...
```

I'll note I don't see any instances of the tangent type being specialised in ChainRules (or in BlockDiagonals)

mzgubic replied:
oh, I see, it's probably this line: https://github.com/FluxML/Zygote.jl/blob/7f2b16987dc4eab236aabd889444cabdcc6f9caf/src/compiler/chainrules.jl#L128

But anyway, I do agree we don't need to parameterise the `Tangent`. I was just wondering whether we should remove the comment (at least in the present form I found it confusing rather than illuminating)

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,63 @@
+    return primal, times_pullback
+end
+
+function ChainRulesCore.rrule(::typeof(*), A::Real, B::WoodburyPDMat, )
```
```suggestion
function ChainRulesCore.rrule(::typeof(*), A::Real, B::WoodburyPDMat)
```

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,63 @@
+    ChainRulesCore.ProjectTo{T}(; fields...)
+end
+
+function (project::ProjectTo{T})(X̄::AbstractMatrix) where {T<:WoodburyPDMat}
```
I don't understand what this does. It looks like we are trying to project an arbitrary matrix onto a Woodbury, which as I understand is not possible in case X is not positive semi-definite? (In that case we might explicitly check for PSD and throw an error?)

If all is well and I just don't understand something, could we add a few lines of comments that explain what is going on and maybe link a reference for the math?

AlexRobson replied:
> trying to project an arbitrary matrix onto a Woodbury

Nope - we are projecting an arbitrary matrix onto a Woodbury _tangent_. I.e. instead of passing around W̄
, we are passing around Ā, D̄, S̄

The matrix here can be anything. Project here is encoding the rules on how the individual components of the woodbury, i.e. from the definition `W = ADA' + S -> dW = ADdA' + AdDA' + dS` => Ā = 2ADW̄, D̄=AW̄A', S̄ = W̄ with their appropriate projectiosn down  to e.g. Diagonal. X̄ here is W̄.

Ideally, these would only be in the constructor for the Woodbury, however, due to the way FD is set-up, it seems we always have to project down to these components. THe woodbury is a lazy low-rank matrix and because of this any woodbury primal has number of elements 2*(A1 + A2) where (A1, A2) is the size of the matrix A. And the tangent has to be the same size (i.e. number of elements), i.e. their `to_vec` representation must equal in length

This is why when testing with a Matrix cotangent I couldn't use CRTU and had to do a manual call to FD and densify the primal [here](https://github.com/invenia/PDMatsExtras.jl/blob/ff8d903bba668e8be6a115866e6abfc1a12d4700/test/chainrules.jl#L35)

If need be, I believe we  can only ever return the natural differential Matrix (i.e. W̄ ) as the outut of any abitrary rrule with a Woodbury argument as is done [here](https://github.com/invenia/PDMatsExtras.jl/blob/ar/chainrules/src/chainrules.jl#L25) but as I say I believe that FD will not like this in some scenarios. 

The constraints imposed here are, as I understand them are that the primal output and output cotangent need to be of the same length, and that the input args and input adjoints are also all of the same length to each other (where length is the to_vec representation). 

I'll add comments, but first I want to make sure that that makes sense

mzgubic replied:
Hmm, I am actually not sure what we should do here, I've opened https://github.com/JuliaDiff/ChainRulesCore.jl/issues/442
with the `PositiveReal` example you have showed me.

AlexRobson replied:
I'm reading that thread. 

The use case here aside, Is projecting onto a `Tangent` not an intended use of `ProjectTo`?
I'm sort of getting that impression from reading these threads, and it is not one that I had realised, if that is the case.

mzgubic replied:
Sure, we can map to `Tangent` because `Tangent` is a differential type. The differences is that
```julia
ProjectTo(w::Tangent{Woodbury})
```
is fine, but
```julia
ProjectTo(w::Woodbury)
```
is not, and we are doing the second here. (Despite the fact that the projection actually returns the `Tangent{Woodbury}`, which is inconsistent since we said we are projecting onto a `Woodbury`)

AlexRobson replied:
Ohhhhhh I see. I'd misunderstood the semantics of `ProjectTo` (and the objection here). 

In which case, I think what I want to do is impossible under the set-up here as it doesn't make sense to keep the woodbury, or in fact any constrained type. 

So any rrule with ProjectTo need to have custom rules when it uses constrained types

But I think this conversation can now fall under the issue you created.

mzgubic replied:
yeah let's see what conclusion we reach there. (I suspect that we will allow this, but let's not merge this until we actually do)

AlexRobson replied:
How does this relate to the proposal in https://github.com/JuliaDiff/ChainRulesCore.jl/pull/449?

mzgubic replied:
I think Will thought that it is possible to find the inverse of Woodbury densification (for the subspace of dense matrices that _can_ come from a Woodbury) but I may have misinterpreted that.

This seems to be a weird edge case @willtebbutt. If I understand correctly, in this case we want the structural rather than the natural differential inside the pullback written in the rule (rather than the premise in the 449 which expects AD to prefer structural tangents, and rule writers to prefer natural tangents)

willtebbutt replied:
The proposal in 449 would basically replace `ProjectTo` / change its semantics to eat a natural tangent / `Matrix` and produce a structural tangent. You obtain the new thing by implementing the pullback for `collect`, and that does everything you need.

Re inverse of densification -- I found that hard to implement for `WoodburyPDMat` -- see [here](https://github.com/JuliaDiff/ChainRulesCore.jl/pull/449/files?file-filters%5B%5D=.jl&file-filters%5B%5D=.md#diff-7b46aec105a00326f97d8740cefcee1c791fb89447f6637410d4855b8b84ffe2R475) but the pullback for `collect` (`pullback_of_destructure` -- I wanted to give it a different name) is totally fine.

willtebbutt replied:
Separately, what's the rationale for implementing `ProjectTo`? Is it to make it possible to do `*(A::Matrix, B::WoodburyPDMat)`? If so, would it make more sense / be easier to implement a specialised method for the primal (since the generic fallback will be horribly slow) that is differentiable, and `@opt_out` of the generic rrule to let AD do the work of deriving the rule?

AlexRobson replied:
> Separately, what's the rationale for implementing ProjectTo

Originally, I was using the `ProjectTo` here to project the natural tangent onto it's structural tangent - i.e. for a given W̄ (the natural tangent, as I understand the semantics), to it's structural (Ā, D̄, S̄), but it turned out this wasn't permitted under the semantics of ProjectTo - which what led to the 'allow projectto onto tangent' [issue ](https://github.com/JuliaDiff/ChainRulesCore.jl/issues/442). I paused looking at this MR at that point, but saw your RFC being posted and pinged this thread on how these two related. As far as I understand the resolution to #442 in CRC it's added an identity, but I don't think that changes anything w.r.t using it as a tangent, so I still wouldn't know how to continue this. 

*If* I understand your RFC, the way I was using the projectto here was more in line with what you were considering.

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,63 @@
+    return (NoTangent(), Ā, B̄)
+end
+
+function _times_pullback(Ȳ::Tangent, A::Real, B::T, proj) where {T<:WoodburyPDMat}
+    Ā = @thunk proj.A(dot(Ȳ.D, B.D) + dot(Ȳ.S, B.S))
+    B̄ = @thunk proj.B(Tangent{WoodburyPDMat}(; A = Ȳ.A, D = Ȳ.D * A, S = Ȳ.S * A))
+    return (NoTangent(), Ā, B̄)
+end
```
could we do something similar to the forward pass? like
```suggestion
function _times_pullback(Ȳ::Tangent, A::Real, B::T, proj) where {T<:WoodburyPDMat}
    df, db, da = _times_pullback(Ȳ, B, A, (A=proj.B, B=proj.A))
    return df, db, da
end
```
(I don't think we have to worry about adjointing a Real)

mzgubic commented on src/chainrules.jl:
```diff
@@ -0,0 +1,63 @@
+    ) where  {T<:WoodburyPDMat}
+    return WoodburyPDMat(A, D, S), X̄ -> WoodburyPDMat_pullback(X̄, A, D, S)
+end
+WoodburyPDMat_pullback(X̄::Tangent, A, D, S) = (X̄ = (unthunk(X̄); return (NoTangent(), X̄.A, X̄.D, X̄.S)))
```
```suggestion
WoodburyPDMat_pullback(X̄::Tangent, A, D, S) = (NoTangent(), X̄.A, X̄.D, X̄.S)
```
we already know it is not a thunk

a separate thing is that this will error if receiving an AbstractMatrix - though i think this is fine, we'll cross that bridge once we get to it

mzgubic commented on test/chainrules.jl:
```diff
@@ -0,0 +1,45 @@
+        )
+    end
+
+    @testset "*(Matrix-Woodbury)" begin
```
I don't see this rule anywhere, does this work because ChainRules are loaded? might be worth adding a comment

AlexRobson replied:
Oh I think these are because Zygote is loaded. 

I'll add ChainRules explicitely

mzgubic commented on test/chainrules.jl:
```diff
@@ -0,0 +1,45 @@
+
+    @testset "Constructors" begin
+        test_rrule(WoodburyPDMat, W.A, W.D, W.S)
+        # This is a gradient. Should be able to deal with negative elements
```
```suggestion
        # This is a gradient, should be able to deal with negative elements (does not have to be PSD like Woodbury itself)
```

willtebbutt commented on src/chainrules.jl:
```diff
@@ -0,0 +1,69 @@
+
+# Rule for Woodbury * Real. 
+# Ignoring Complex version for now. 
+function ChainRulesCore.rrule(::typeof(*), A::WoodburyPDMat, B::Real)
```
Sorry to hop in late here, but why do we need a rule for this, as opposed to just opting out?

In particular, our implementation of this function ought to be straightforwardly differentiable by Zygote et al:
```julia
*(a::WoodburyPDMat, c::Real) = WoodburyPDMat(a.A, a.D * c, a.S * c)
*(c::Real, a::WoodburyPDMat) = a * c
```

Is it not sufficient just to `@opt_out`?

AlexRobson replied:
Yeah this is probabaly true. AFAIK this works - here however I was using this to test where the primal returns a woodbury (and thus your tangent is woodbury) as I wanted to use this as a basis to sort out how to work on the rule for `*(H::Diagonal, S::WoodburyPDMat)`. If you remember we had issues with that and had to do `WoodburyPDMat(H * S.A, S.D, H * S.S * H' + D)`. This was actually the direction I wanted to take this MR, but  possibly a lot has moved on since that issue. 

I hadn't considered opt-outs... if the rule works are they generally not more efficient than zygote working it out?

willtebbutt replied:
> Yeah this is probabaly true. AFAIK this works - here however I was using this to test where the primal returns a woodbury (and thus your output co-tangent is a tangent{woodbury}) as I wanted to use this as a basis to sort out how to work on the rule for *(H::Diagonal, S::WoodburyPDMat). If you remember we had issues with that and had to do WoodburyPDMat(H * S.A, S.D, H * S.S * H' + D). This was actually the direction I wanted to take this MR, but possibly a lot has moved on since that issue.

Interesting -- again, I'd suggest just opting out here 🤷 Or am I missing the point?

> I hadn't considered opt-outs... if the rule works are they generally not more efficient than zygote working it out?

No. In simple cases like this there's no reason that Zygote can't achieve optimal performance. If you have control flow (either via a for-loop or an `if-else-end` block) Zygote will generally struggle, but otherwise you're definitely better off letting Zygote figure stuff out.

My experience has consistently been that if you have the choice between writing a rule and writing AD-friendly code, always pick the latter.

willtebbutt reviewed (commented):
Sorry for jumping in late. Saw this linked in 449.

AlexRobson commented:
I'm just going to close this MR. I had largely set-out to try and understand CR through implementing rules for the woodbury (and follow up with a resolution to the Woodbury * diagonal AD workaround that was neccessary) with the presumption that the rules would be more performant (i neever got to benchmarking these) but this seems the wrong view. 

There may be some follow-ups (such as opt-outs and supporting CRC 1) but it seems best to do that with a clean slate.
